include/gridtools/storage/adapter/nanobind_adapter.hpp

```diff
@@ -0,0 +1,86 @@
+        };
+
+        template <std::size_t N>
+        using fully_dynamic_strides = typename dynamic_strides_helper<std::make_index_sequence<N>>::type;
+
+        template <std::size_t SpecValue>
+        auto select_static_stride_value(std::size_t dyn_value) {
+            if constexpr (SpecValue == nanobind::any) {
+                return dyn_value;
+            } else {
+                if (SpecValue != dyn_value) {
+                    throw std::invalid_argument("static stride in stride specification doesn't match dynamic stride");
+                }
+                return gridtools::integral_constant<std::size_t, SpecValue>{};
+            }
+        }
+
+        template <std::size_t... SpecValues, std::size_t... IndexValues>
+        auto select_static_strides_helper(
+            stride_spec<SpecValues...>, const std::size_t *dyn_values, std::index_sequence<IndexValues...>) {
+
+            return gridtools::tuple{select_static_stride_value<SpecValues>(dyn_values[IndexValues])...};
+        }
+
+        template <std::size_t... SpecValues>
+        auto select_static_strides(stride_spec<SpecValues...> spec, const std::size_t *dyn_values) {
+            return select_static_strides_helper(spec, dyn_values, std::make_index_sequence<sizeof...(SpecValues)>{});
+        }
+
+        template <class T,
+            std::size_t... Sizes,
+            class... Args,
+            class Strides = fully_dynamic_strides<sizeof...(Sizes)>,
+            class StridesKind = sid::unknown_kind>
+        auto as_sid(nanobind::ndarray<T, nanobind::shape<Sizes...>, Args...> ndarray,
+            Strides stride_spec_ = fully_dynamic_strides<sizeof...(Sizes)>{},
+            [[maybe_unused]] StridesKind strides_kind = sid::unknown_kind{}) {
+            using sid::property;
+            const auto ptr = ndarray.data();
+            constexpr auto ndim = sizeof...(Sizes);
+            assert(ndim == ndarray.ndim());
+            gridtools::array<std::size_t, ndim> shape;
+            std::copy_n(ndarray.shape_ptr(), ndim, shape.begin());
+            gridtools::array<std::size_t, ndim> strides_;
+            std::copy_n(ndarray.stride_ptr(), ndim, strides_.begin());
+            const auto strides = select_static_strides(stride_spec_, strides_.data());
+
+            return sid::synthetic()
+                .template set<property::origin>(sid::host_device::simple_ptr_holder<T *>{ptr})
+                .template set<property::strides>(strides)
+                .template set<property::strides_kind, StridesKind>()
+                .template set<property::lower_bounds>(gridtools::array<integral_constant<std::size_t, 0>, ndim>())
+                .template set<property::upper_bounds>(shape);
+        }
+    } // namespace nanobind_sid_adapter_impl_
+
+    using nanobind_sid_adapter_impl_::as_sid;
+    using nanobind_sid_adapter_impl_::fully_dynamic_strides;
+    using nanobind_sid_adapter_impl_::stride_spec;
```

Comment:
I don't like exporting these relatively generic names but very specific meaning.
Maybe we should wrap these two functions into `gridtools::nanobind`.

Reply:
Or maybe we don't need `fully_dynamic_strides` to be public, because we can do `as_sid(ndarray, {}, kind)` and maybe `stride_spec` can just be a tuple.

Reply:
Moved them to `gridtools::nanobind`.